standards/bulk.md

```diff
@@ -0,0 +1,174 @@
+- Operation collections in a request **MAY** include entities of the same ID, in which case the operations are performed in the order they are received. In this case the use of `ETag` may not work as expected for subsequent concurrency checks. Operation ID should be used by consumer in this case to identify the outcome of each operation, especially if transaction type is `ISOLATED`.
+
+#### Response
+
+- Bulk operations **MUST** include a response body schema as follows, which is not extensible:
+```json
+{
+    "status": "SUCCEEDED" | "FAILED" | "PARTIAL",                           // overall status of bulk operation, partial indicating 
+    "operations": [                                                         // there are some operations that failed and succeeded
+        {
+            "operationId": "string",                                        // matching operation id from the request body if provided, otherwise use index value as a string
+            "action": "CREATE" | "UPDATE" | "CREATE_UPDATE" | "DELETE",     // repeat action type
+            "entityId": "string" | null,                                    // the associated id of the entity, if available
+            "entityRef": "sps-ref" | null,                                  // the associated sps-ref URN entity, if applicable
+            "status": "SUCCEEDED" | "FAILED",                               // status of individual operation
+            "detail": {                                                     // optional detail information about the operation, like error details.
+                "message": "string",                                        // if detail object is provided, it must include a message
+                "code": "string" | null,                                    // codes can be similar to those used in Error Response, or custom for other purposes
+                "field": "string" | null,                                   // field indicates an associated field in the entity to highlight
+                "value": "string" | null                                    // the value of the associated field highlighted in the detail
+            }
+        }
+    ]
+}
+```
+
+- `status` enumeration **MUST NOT** be extended or filtered.
+- `operations` collection **SHOULD** be in the same order as the request body operations.
+- `operationId` in the response **MUST** match the `operationId` of the request body where provided, otherwise it should fallback to the index of the operation in the request body as a string value.
+- `entityId` and `entityRef` should both be used to identify the primary ID or an associated [ref](naming.md) value for the entity. These are optional and can be left out if not applicable.
+- `detail` field **MAY** be included as a complex object following the identified schema. If it is include, then it **MUST** include a `message` field. It **MUST NOT** be extended.
```

Comment:
In our error response schema, isn't this part a list as there might be multiple validation errors each with their own error code indicating a specific problem? It looks like we have a single object here. Maybe that's on purpose to limit some complexity, but that means errors would need to be reported differently here than if the entity was acted on individually through a non-bulk endpoint.

Additionally, I thought `field` and `value` were example extended error context fields that weren't necessarily required or exhaustive. The way it's presented here is that they'd be mandatory in this bulk schema that is also deemed not extensible. Seems like this is a bit inconsistent too?

Reply:
Your suspicions was where my head was at writing this:
- Leaving it as a single detail object was intended to reduce complexity, especially in scenarios in my head where much of the validation of handling a single resource would be a single object. So you end up with many usages where detail is a collection of one.
- I expected that there were additional validation concerns that happen when using a resource directly around headers and other request validation issues that often resulted in validation having multiple messages / codes. Typically I'd bet we would find a non-exhaustive return of all possible validations, usually just being the first one that we fail on. Again, with that in mind, it was hard for me to get to making detail a collection all the time.
- field and error are examples of error extension, but enough of an example codified in a domain that if needed teams will copy that pattern. Here it is the same, though not extendable in order to lock it down until we have a real use case where something different is needed. I did make a comment though that if detail is provided, message must be provided, but the rest of the fields were optional on the resource as a whole (i.e. not dynamic schema for each operation). 

The duality you speak of between error validations for a 400 problem details and here are still rattling in my brain though. I'm wondering if if this should be specifically error context to tie together the same schema and collection of validation messages. A separate top string message field as a sibling to error context could provide the regular use case message (allowing error collection also to be null when no errors occurred). However, I had considered detail previously to be usable by both success and failure... for example to indicate the value of a new field on success - that might be way to theoritical though.

```json
{
    "status": "SUCCEEDED" | "FAILED" | "PARTIAL",                           
    "operations": [                                                         
        {
            "operationId": "string",                                        
            "action": "CREATE" | "UPDATE" | "CREATE_UPDATE" | "DELETE",     
            "entityId": "string" | null,                                    
            "entityRef": "sps-ref" | null,                                  
            "status": "SUCCEEDED" | "FAILED", 
            "result": {                                     // Details pertaining to the result of the operation. -> DO WE NEED THIS OR JUST PUT DETAIL/CONTEXT at ROOT of operation object?
                "detail": "string" | null,                  // Description or detailed human-readable message about the success or failure of the operation.            
                "context": [                                // OPTIONAL (array): List of objects providing additional context and detail on sub-reasons for any errors or failures.
                    {
                        "message": "string",                // REQUIRED (string): Human-readable details or message specific error about the request failure.                       
                        "code": "string" | null,            // OPTIONAL (string): Short, machine-readable, name of the validation error that occurred.                        
                        "field": "string" | null,                                   
                        "value": "string" | null 
                    }                                   
                ]
            }
        }
    ]
}
```

Thoughts?

Reply:
Yeah, I like the general idea/direction of that. The complexity is really around handling errors, not success, so that might deserve its own object that is just only used when there was an error. And since we already have a concept for that, I'd think we'd want some consistency. Curious what other folks think, though.

`status` could maybe move under `result` too since it's part of the output and not identifying the operation or request. But `status` does also feel like it should be a top-level thing for each operation where it is now. 🤔 

Or if we don't want to nest things as much, `detail` and `context` could probably be flatter on the object as you mentioned in comments - but maybe rename them to fit there a bit better or something?

Reply:
I like status in result. And while it is convenient at the root, I think having it in result is probably better of the two for grouping. Asking around a bit, there was a desire from others to keep context and detail for generic usage. The result of the operation itself as a success/failure isn't inherently a concern of the contract, its just about displaying the results.

Take a look at the update and let me know if you have strong opinions in either direction?